asapdiscovery-simulation/asapdiscovery/simulation/schema/fec.py

```diff
@@ -0,0 +1,378 @@
+    """
+    Store the results of a transformation, note when retries are used this will be the average result.
+    """
+
+    type: Literal["TransformationResult"] = "TransformationResult"
+    ligand_a: str = Field(
+        ..., description="The name of the ligand in state A of the transformation."
+    )
+    ligand_b: str = Field(
+        ..., description="The name of the ligand in state B of the transformation."
+    )
+    phase: Literal["complex", "solvent"] = Field(
+        ..., description="The phase of the transformation."
+    )
+    estimate: FloatQuantity["kcal/mol"] = Field(  # noqa: F821
+        ..., description="The average estimate of this transformation in kcal/mol"
+    )
+    uncertainty: FloatQuantity["kcal/mol"] = Field(  # noqa: F821
+        ...,
+        description="The standard deviation of the estimates of this transform in kcal/mol",
+    )
+
+    def name(self):
+        """Make a name for this transformation based on the names of the ligands."""
+        return "-".join([self.ligand_a, self.ligand_b])
+
+
+class _BaseResults(_SchemaBaseFrozen):
+    """
+    A base results class which handles the collecting and processing of the results.
+    """
+
+    type: Literal["_BaseResults"] = "_BaseResults"
+    results: list[TransformationResult] = Field(
+        [], description="The list of results collected for this dataset."
+    )
+
+    def to_cinnabar_measurements(self):
+        """
+        For the given set of results combine the solvent and complex phases to make a list of cinnabar RelativeMeasurements
+
+        Returns:
+            A list of RelativeMeasurements made from the combined solvent and complex phases.
+        """
+        from collections import defaultdict
+
+        import numpy as np
+        from cinnabar import RelativeMeasurement
+
+        raw_results = defaultdict(list)
+        # gather by transform
+        for result in self.results:
+            raw_results[result.name()].append(result)
+
+        # make sure we have a solvent and complex phase for each result
+        for name, transforms in raw_results.items():
+            missing_phase = {"complex", "solvent"} - {t.phase for t in transforms}
+            if missing_phase:
+                raise RuntimeError(
+                    f"The transformation {name} is missing simulated legs in the following phases {missing_phase}"
+                )
+
+        # make the cinnabar data
+        all_results = []
+        for leg1, leg2 in raw_results.values():
+            complex_leg: TransformationResult = (
+                leg1 if leg1.phase == "complex" else leg2
+            )
+            solvent_leg: TransformationResult = (
+                leg1 if leg1.phase == "solvent" else leg2
+            )
+            result = RelativeMeasurement(
+                labelA=leg1.ligand_a,
+                labelB=leg1.ligand_b,
+                DDG=(complex_leg.estimate - solvent_leg.estimate),
+                uncertainty=np.sqrt(
```

Comment:
Make a function or  a note that this is for error propagation, may be useful elsewhere or at least to note to someone following along how its done.